cpp/example/simple_kv_store.cc

```diff
@@ -0,0 +1,207 @@
+
+// Name of the main server actor.
+const std::string MAIN_SERVER_NAME = "main_actor";
+// Name of the backup server actor.
+const std::string BACKUP_SERVER_NAME = "backup_actor";
+// Resources that each actor needs: 1 CPU core and 1 GB memory. 
+const std::unordered_map<std::string, double> RESOUECES{
+    {"CPU", 1.0}, {"memory", 1024.0 * 1024.0 * 1024.0}};
+
+namespace common {
+inline std::pair<bool, std::string> Get(
+    const std::string &key, const std::unordered_map<std::string, std::string> &data) {
+  auto it = data.find(key);
+  if (it == data.end()) {
+    return std::pair<bool, std::string>{};
+  }
+
+  return {true, it->second};
+}
+}  // namespace common
+
+class MainServer;
+class BackupServer {
+ public:
+  BackupServer();
+
+  // The main server will get all BackupServer's data when it restarts.
+  std::unordered_map<std::string, std::string> GetAllData();
+
+  // This is used for the main server to backup data before putting data.
+  void BackupData(const std::string &key, const std::string &val);
+
+ private:
+  // When the backup server restarts from a failure. It will get all data from the main
+  // server.
+  void HandleFailover();
+  // A map that stores the key-value data.
+  std::unordered_map<std::string, std::string> data_;
+};
+
+BackupServer::BackupServer() {
+  // Handle failover when the actor restarts.
+  if (ray::WasCurrentActorRestarted()) {
+    HandleFailover();
+  }
+  RAYLOG(INFO) << "BackupServer created";
+}
+
+std::unordered_map<std::string, std::string> BackupServer::GetAllData() { return data_; }
+
+void BackupServer::BackupData(const std::string &key, const std::string &val) {
+  data_[key] = val;
+}
+
+class MainServer {
+ public:
+  MainServer();
+
+  // Set value by the given key.
+  void Put(const std::string &key, const std::string &val);
+
+  // Get value from a key.
+  std::pair<bool, std::string> Get(const std::string &key);
+
+  // This is used for the backup server to recover data when it restarts.
+  std::unordered_map<std::string, std::string> GetAllData();
+
+ private:
+  // When the main server restarts from a failure. It will get all data from the backup
+  // server.
+  void HandleFailover();
+  // A map that stores the key-value data.
+  std::unordered_map<std::string, std::string> data_;
+  // Actor handle to the backup server actor.
+  ray::ActorHandle<BackupServer> backup_actor_;
+};
+
+MainServer::MainServer() {
+  if (ray::WasCurrentActorRestarted()) {
+    HandleFailover();
+  }
+
+  backup_actor_ = *ray::GetActor<BackupServer>(BACKUP_SERVER_NAME);
+  RAYLOG(INFO) << "MainServer created";
+}
+
+void MainServer::Put(const std::string &key, const std::string &val) {
+  // Backup data before putting data.
+  auto r = backup_actor_.Task(&BackupServer::BackupData).Remote(key, val);
+  std::vector<ray::ObjectRef<void>> objects{r};
+  auto result = ray::Wait(objects, objects.size(), 2000);
+  if (result.ready.empty()) {
+    RAYLOG(WARNING) << "MainServer BackupData failed.";
+  }
```

Comment:
just `backup_actor_.Task(&BackupServer::BackupData).Remote(key, val).Get()`?

Reply:
I have used Ray Get in client, it will be never timeout, here i want to show how to wait for object with timeout.